File: src/Solutions/Euler125.hs
module Solutions.Euler125 where
import Utils.List(uniques)
import Utils.Numeric(isPalindrome)
-- sum of squares from 1 to n
sumSquares n = n * (n + 1) * (2 * n + 1) `div` 6

compute :: Int
compute = sum $ uniques [k |
    a <- [2..10000],
    b <- [0..a - 2],
    let k = sumSquares a - sumSquares b,
    isPalindrome k,
    k < 10^8]

euler125 :: IO String
euler125 = return $ show compute 

File: rust/src/solutions/euler125.rs
use crate::utils::numeric::digits::Digits;

pub fn euler125() -> String {
    let mut palindromic: Vec<u64> = Vec::new();
    let sum_of_squares: Vec<u64> = (0..=10000).map(|n| n * (n + 1) * (2 * n + 1) / 6).collect();
    for a in 2..=10000 {
        for b in 0..a - 1 {
            let k = sum_of_squares[a] - sum_of_squares[b];
            if k.is_palindrome() && k < 100_000_000 {
                palindromic.push(k);
            }
        }
    }
    palindromic.sort();
    palindromic.dedup();
    palindromic.into_iter().sum::<u64>().to_string()
}
